File: src/extractors/csharp.js
import { extractModifierVisibility, findChild, nodeEndLine } from './helpers.js';

/**
 * Extract symbols from C# files.
 */
export function extractCSharpSymbols(tree, _filePath) {
  const ctx = {
    definitions: [],
    calls: [],
    imports: [],
    classes: [],
    exports: [],
    typeMap: new Map(),
  };

  walkCSharpNode(tree.rootNode, ctx);
  extractCSharpTypeMap(tree.rootNode, ctx);
  return ctx;
}

function walkCSharpNode(node, ctx) {
  switch (node.type) {
    case 'class_declaration':
      handleCsClassDecl(node, ctx);
      break;
    case 'struct_declaration':
      handleCsStructDecl(node, ctx);
      break;
    case 'record_declaration':
      handleCsRecordDecl(node, ctx);
      break;
    case 'interface_declaration':
      handleCsInterfaceDecl(node, ctx);
      break;
    case 'enum_declaration':
      handleCsEnumDecl(node, ctx);
      break;
    case 'method_declaration':
      handleCsMethodDecl(node, ctx);
      break;
    case 'constructor_declaration':
      handleCsConstructorDecl(node, ctx);
      break;
    case 'property_declaration':
      handleCsPropertyDecl(node, ctx);
      break;
    case 'using_directive':
      handleCsUsingDirective(node, ctx);
      break;
    case 'invocation_expression':
      handleCsInvocationExpr(node, ctx);
      break;
    case 'object_creation_expression':
      handleCsObjectCreation(node, ctx);
      break;
  }

  for (let i = 0; i < node.childCount; i++) walkCSharpNode(node.child(i), ctx);
}

// ── Walk-path per-node-type handlers ────────────────────────────────────────

function handleCsClassDecl(node, ctx) {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const classChildren = extractCSharpClassFields(node);
  ctx.definitions.push({
    name: nameNode.text,
    kind: 'class',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
    children: classChildren.length > 0 ? classChildren : undefined,
  });
  extractCSharpBaseTypes(node, nameNode.text, ctx.classes);
}

function handleCsStructDecl(node, ctx) {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const structChildren = extractCSharpClassFields(node);
  ctx.definitions.push({
    name: nameNode.text,
    kind: 'struct',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
    children: structChildren.length > 0 ? structChildren : undefined,
  });
  extractCSharpBaseTypes(node, nameNode.text, ctx.classes);
}

function handleCsRecordDecl(node, ctx) {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  ctx.definitions.push({
    name: nameNode.text,
    kind: 'record',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
  });
  extractCSharpBaseTypes(node, nameNode.text, ctx.classes);
}

function handleCsInterfaceDecl(node, ctx) {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  ctx.definitions.push({
    name: nameNode.text,
    kind: 'interface',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
  });
  const body = node.childForFieldName('body');
  if (body) {
    for (let i = 0; i < body.childCount; i++) {
      const child = body.child(i);
      if (child && child.type === 'method_declaration') {
        const methName = child.childForFieldName('name');
        if (methName) {
          ctx.definitions.push({
            name: `${nameNode.text}.${methName.text}`,
            kind: 'method',
            line: child.startPosition.row + 1,
            endLine: child.endPosition.row + 1,
          });
        }
      }
    }
  }
}

function handleCsEnumDecl(node, ctx) {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const enumChildren = extractCSharpEnumMembers(node);
  ctx.definitions.push({
    name: nameNode.text,
    kind: 'enum',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
    children: enumChildren.length > 0 ? enumChildren : undefined,
  });
}

function handleCsMethodDecl(node, ctx) {
  // Skip interface methods already emitted by handleCsInterfaceDecl
  if (node.parent?.parent?.type === 'interface_declaration') return;
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const parentType = findCSharpParentType(node);
  const fullName = parentType ? `${parentType}.${nameNode.text}` : nameNode.text;
  const params = extractCSharpParameters(node.childForFieldName('parameters'));
  ctx.definitions.push({
    name: fullName,
    kind: 'method',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
    children: params.length > 0 ? params : undefined,
    visibility: extractModifierVisibility(node),
  });
}

function handleCsConstructorDecl(node, ctx) {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const parentType = findCSharpParentType(node);
  const fullName = parentType ? `${parentType}.${nameNode.text}` : nameNode.text;
  const params = extractCSharpParameters(node.childForFieldName('parameters'));
  ctx.definitions.push({
    name: fullName,
    kind: 'method',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
    children: params.length > 0 ? params : undefined,
    visibility: extractModifierVisibility(node),
  });
}

function handleCsPropertyDecl(node, ctx) {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const parentType = findCSharpParentType(node);
  const fullName = parentType ? `${parentType}.${nameNode.text}` : nameNode.text;
  ctx.definitions.push({
    name: fullName,
    kind: 'property',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
    visibility: extractModifierVisibility(node),
  });
}

function handleCsUsingDirective(node, ctx) {
  const nameNode =
    node.childForFieldName('name') ||
    findChild(node, 'qualified_name') ||
    findChild(node, 'identifier');
  if (!nameNode) return;
  const fullPath = nameNode.text;
  const lastName = fullPath.split('.').pop();
  ctx.imports.push({
    source: fullPath,
    names: [lastName],
    line: node.startPosition.row + 1,
    csharpUsing: true,
  });
}

function handleCsInvocationExpr(node, ctx) {
  const fn = node.childForFieldName('function') || node.child(0);
  if (!fn) return;
  if (fn.type === 'identifier') {
    ctx.calls.push({ name: fn.text, line: node.startPosition.row + 1 });
  } else if (fn.type === 'member_access_expression') {
    const name = fn.childForFieldName('name');
    if (name) {
      const expr = fn.childForFieldName('expression');
      const call = { name: name.text, line: node.startPosition.row + 1 };
      if (expr) call.receiver = expr.text;
      ctx.calls.push(call);
    }
  } else if (fn.type === 'generic_name' || fn.type === 'member_binding_expression') {
    const name = fn.childForFieldName('name') || fn.child(0);
    if (name) ctx.calls.push({ name: name.text, line: node.startPosition.row + 1 });
  }
}

function handleCsObjectCreation(node, ctx) {
  const typeNode = node.childForFieldName('type');
  if (!typeNode) return;
  const typeName =
    typeNode.type === 'generic_name'
      ? typeNode.childForFieldName('name')?.text || typeNode.child(0)?.text
      : typeNode.text;
  if (typeName) ctx.calls.push({ name: typeName, line: node.startPosition.row + 1 });
}

function findCSharpParentType(node) {
  let current = node.parent;
  while (current) {
    if (
      current.type === 'class_declaration' ||
      current.type === 'struct_declaration' ||
      current.type === 'interface_declaration' ||
      current.type === 'enum_declaration' ||
      current.type === 'record_declaration'
    ) {
      const nameNode = current.childForFieldName('name');
      return nameNode ? nameNode.text : null;
    }
    current = current.parent;
  }
  return null;
}

// ── Child extraction helpers ────────────────────────────────────────────────

function extractCSharpParameters(paramListNode) {
  const params = [];
  if (!paramListNode) return params;
  for (let i = 0; i < paramListNode.childCount; i++) {
    const param = paramListNode.child(i);
    if (!param || param.type !== 'parameter') continue;
    const nameNode = param.childForFieldName('name');
    if (nameNode) {
      params.push({ name: nameNode.text, kind: 'parameter', line: param.startPosition.row + 1 });
    }
  }
  return params;
}

function extractCSharpClassFields(classNode) {
  const fields = [];
  const body = classNode.childForFieldName('body') || findChild(classNode, 'declaration_list');
  if (!body) return fields;
  for (let i = 0; i < body.childCount; i++) {
    const member = body.child(i);
    if (!member || member.type !== 'field_declaration') continue;
    const varDecl = findChild(member, 'variable_declaration');
    if (!varDecl) continue;
    for (let j = 0; j < varDecl.childCount; j++) {
      const child = varDecl.child(j);
      if (!child || child.type !== 'variable_declarator') continue;
      const nameNode = child.childForFieldName('name');
      if (nameNode) {
        fields.push({
          name: nameNode.text,
          kind: 'property',
          line: member.startPosition.row + 1,
          visibility: extractModifierVisibility(member),
        });
      }
    }
  }
  return fields;
}

function extractCSharpEnumMembers(enumNode) {
  const constants = [];
  const body =
    enumNode.childForFieldName('body') || findChild(enumNode, 'enum_member_declaration_list');
  if (!body) return constants;
  for (let i = 0; i < body.childCount; i++) {
    const member = body.child(i);
    if (!member || member.type !== 'enum_member_declaration') continue;
    const nameNode = member.childForFieldName('name');
    if (nameNode) {
      constants.push({ name: nameNode.text, kind: 'constant', line: member.startPosition.row + 1 });
    }
  }
  return constants;
}

// ── Type map extraction ──────────────────────────────────────────────────────

function extractCSharpTypeMap(node, ctx) {
  extractCSharpTypeMapDepth(node, ctx, 0);
}

function extractCSharpTypeMapDepth(node, ctx, depth) {
  if (depth >= 200) return;

  // local_declaration_statement → variable_declaration → type + variable_declarator(s)
  if (node.type === 'variable_declaration') {
    const typeNode = node.childForFieldName('type') || node.child(0);
    if (typeNode && typeNode.type !== 'var_keyword') {
      const typeName = extractCSharpTypeName(typeNode);
      if (typeName) {
        for (let i = 0; i < node.childCount; i++) {
          const child = node.child(i);
          if (child && child.type === 'variable_declarator') {
            const nameNode = child.childForFieldName('name') || child.child(0);
            if (nameNode && nameNode.type === 'identifier') {
              ctx.typeMap.set(nameNode.text, { type: typeName, confidence: 0.9 });
            }
          }
        }
      }
    }
  }

  // Method/constructor parameter: parameter node has type + name fields
  if (node.type === 'parameter') {
    const typeNode = node.childForFieldName('type');
    const nameNode = node.childForFieldName('name');
    if (typeNode && nameNode) {
      const typeName = extractCSharpTypeName(typeNode);
      if (typeName) ctx.typeMap.set(nameNode.text, { type: typeName, confidence: 0.9 });
    }
  }

  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child) extractCSharpTypeMapDepth(child, ctx, depth + 1);
  }
}

function extractCSharpTypeName(typeNode) {
  if (!typeNode) return null;
  const t = typeNode.type;
  if (t === 'identifier' || t === 'qualified_name') return typeNode.text;
  if (t === 'predefined_type') return null; // skip int, string, etc
  if (t === 'generic_name') {
    const first = typeNode.child(0);
    return first ? first.text : null;
  }
  if (t === 'nullable_type') {
    const inner = typeNode.child(0);
    return inner ? extractCSharpTypeName(inner) : null;
  }
  return null;
}

function extractCSharpBaseTypes(node, className, classes) {
  const baseList = node.childForFieldName('bases');
  if (!baseList) return;
  for (let i = 0; i < baseList.childCount; i++) {
    const child = baseList.child(i);
    if (!child) continue;
    if (child.type === 'identifier' || child.type === 'qualified_name') {
      classes.push({ name: className, extends: child.text, line: node.startPosition.row + 1 });
    } else if (child.type === 'generic_name') {
      const name = child.childForFieldName('name') || child.child(0);
      if (name)
        classes.push({ name: className, extends: name.text, line: node.startPosition.row + 1 });
    } else if (child.type === 'base_list') {
      for (let j = 0; j < child.childCount; j++) {
        const base = child.child(j);
        if (base && (base.type === 'identifier' || base.type === 'qualified_name')) {
          classes.push({ name: className, extends: base.text, line: node.startPosition.row + 1 });
        } else if (base && base.type === 'generic_name') {
          const name = base.childForFieldName('name') || base.child(0);
          if (name)
            classes.push({ name: className, extends: name.text, line: node.startPosition.row + 1 });
        }
      }
    }
  }
}

File: crates/codegraph-core/src/extractors/csharp.rs
use super::helpers::*;
use super::SymbolExtractor;
use crate::ast_analysis::cfg::build_function_cfg;
use crate::ast_analysis::complexity::compute_all_metrics;
use crate::types::*;
use std::collections::HashSet;
use tree_sitter::{Node, Tree};

pub struct CSharpExtractor;

impl SymbolExtractor for CSharpExtractor {
    fn extract(&self, tree: &Tree, source: &[u8], file_path: &str) -> FileSymbols {
        let mut symbols = FileSymbols::new(file_path.to_string());
        walk_tree(&tree.root_node(), source, &mut symbols, match_csharp_node);
        reclassify_csharp_implements(&mut symbols);
        walk_ast_nodes_with_config(&tree.root_node(), source, &mut symbols.ast_nodes, &CSHARP_AST_CONFIG);
        walk_tree(&tree.root_node(), source, &mut symbols, match_csharp_type_map);
        dedup_type_map(&mut symbols.type_map);
        symbols
    }
}

const CSHARP_TYPE_KINDS: &[&str] = &[
    "class_declaration", "struct_declaration", "interface_declaration",
    "enum_declaration", "record_declaration",
];

fn find_csharp_parent_type(node: &Node, source: &[u8]) -> Option<String> {
    find_enclosing_type_name(node, CSHARP_TYPE_KINDS, source)
}

fn match_csharp_node(node: &Node, source: &[u8], symbols: &mut FileSymbols, _depth: usize) {
    match node.kind() {
        "class_declaration" => handle_class_decl(node, source, symbols),
        "struct_declaration" => handle_struct_decl(node, source, symbols),
        "record_declaration" => handle_record_decl(node, source, symbols),
        "interface_declaration" => handle_interface_decl(node, source, symbols),
        "enum_declaration" => handle_enum_decl(node, source, symbols),
        "method_declaration" => handle_method_decl(node, source, symbols),
        "constructor_declaration" => handle_constructor_decl(node, source, symbols),
        "property_declaration" => handle_property_decl(node, source, symbols),
        "using_directive" => handle_using_directive(node, source, symbols),
        "invocation_expression" => handle_invocation_expr(node, source, symbols),
        "object_creation_expression" => handle_object_creation(node, source, symbols),
        _ => {}
    }
}

// ── Per-node-kind handlers for walk_node_depth ───────────────────────────────

fn handle_class_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(name_node) = node.child_by_field_name("name") else { return };
    let class_name = node_text(&name_node, source).to_string();
    let children = extract_csharp_class_fields(node, source);
    symbols.definitions.push(Definition {
        name: class_name.clone(),
        kind: "class".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: opt_children(children),
    });
    extract_csharp_base_types(node, &class_name, source, symbols);
}

fn handle_struct_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(name_node) = node.child_by_field_name("name") else { return };
    let name = node_text(&name_node, source).to_string();
    symbols.definitions.push(Definition {
        name: name.clone(),
        kind: "struct".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
    extract_csharp_base_types(node, &name, source, symbols);
}

fn handle_record_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(name_node) = node.child_by_field_name("name") else { return };
    let name = node_text(&name_node, source).to_string();
    symbols.definitions.push(Definition {
        name: name.clone(),
        kind: "record".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
    extract_csharp_base_types(node, &name, source, symbols);
}

fn handle_interface_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(name_node) = node.child_by_field_name("name") else { return };
    let iface_name = node_text(&name_node, source).to_string();
    symbols.definitions.push(Definition {
        name: iface_name.clone(),
        kind: "interface".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
    if let Some(body) = node.child_by_field_name("body") {
        for i in 0..body.child_count() {
            let Some(child) = body.child(i) else { continue };
            if child.kind() != "method_declaration" { continue; }
            if let Some(meth_name) = child.child_by_field_name("name") {
                // Interface method declarations have no body — skip CFG and complexity
                // to mirror the WASM extractor and avoid producing meaningless metrics
                // for body-less declarations.
                symbols.definitions.push(Definition {
                    name: format!("{}.{}", iface_name, node_text(&meth_name, source)),
                    kind: "method".to_string(),
                    line: start_line(&child),
                    end_line: Some(end_line(&child)),
                    decorators: None,
                    complexity: None,
                    cfg: None,
                    children: None,
                });
            }
        }
    }
}

fn handle_enum_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    if let Some(name_node) = node.child_by_field_name("name") {
        let enum_name = node_text(&name_node, source).to_string();
        let children = extract_csharp_enum_members(node, source);
        symbols.definitions.push(Definition {
            name: enum_name,
            kind: "enum".to_string(),
            line: start_line(node),
            end_line: Some(end_line(node)),
            decorators: None,
            complexity: None,
            cfg: None,
            children: opt_children(children),
        });
    }
}

fn handle_method_or_ctor(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(name_node) = node.child_by_field_name("name") else { return };
    let parent_type = find_csharp_parent_type(node, source);
    let name = node_text(&name_node, source);
    let full_name = match &parent_type {
        Some(pt) => format!("{}.{}", pt, name),
        None => name.to_string(),
    };
    let children = extract_csharp_parameters(node, source);
    symbols.definitions.push(Definition {
        name: full_name,
        kind: "method".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: compute_all_metrics(node, source, "csharp"),
        cfg: build_function_cfg(node, "csharp", source),
        children: opt_children(children),
    });
}

fn handle_method_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    // Skip interface methods — already emitted by handle_interface_decl without
    // parameters. Re-emitting them here would duplicate the definition and add
    // spurious `contains` edges to parameters of body-less declarations.
    if let Some(parent) = node.parent() {
        if let Some(grand) = parent.parent() {
            if grand.kind() == "interface_declaration" { return; }
        }
    }
    handle_method_or_ctor(node, source, symbols);
}

fn handle_constructor_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    handle_method_or_ctor(node, source, symbols);
}

fn handle_property_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(name_node) = node.child_by_field_name("name") else { return };
    let parent_type = find_csharp_parent_type(node, source);
    let name = node_text(&name_node, source);
    let full_name = match &parent_type {
        Some(pt) => format!("{}.{}", pt, name),
        None => name.to_string(),
    };
    symbols.definitions.push(Definition {
        name: full_name,
        kind: "property".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
}

fn handle_using_directive(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let name_node = node
        .child_by_field_name("name")
        .or_else(|| find_child(node, "qualified_name"))
        .or_else(|| find_child(node, "identifier"));
    if let Some(name_node) = name_node {
        let full_path = node_text(&name_node, source).to_string();
        let last_name = full_path.split('.').last().unwrap_or("").to_string();
        let mut imp = Import::new(full_path, vec![last_name], start_line(node));
        imp.csharp_using = Some(true);
        symbols.imports.push(imp);
    }
}

fn handle_invocation_expr(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let fn_node = node.child_by_field_name("function").or_else(|| node.child(0));
    let Some(fn_node) = fn_node else { return };
    match fn_node.kind() {
        "identifier" => {
            symbols.calls.push(Call {
                name: node_text(&fn_node, source).to_string(),
                line: start_line(node),
                dynamic: None,
                receiver: None,
            });
        }
        "member_access_expression" => {
            if let Some(name) = fn_node.child_by_field_name("name") {
                let receiver = named_child_text(&fn_node, "expression", source)
                    .map(|s| s.to_string());
                symbols.calls.push(Call {
                    name: node_text(&name, source).to_string(),
                    line: start_line(node),
                    dynamic: None,
                    receiver,
                });
            }
        }
        "generic_name" | "member_binding_expression" => {
            let name = fn_node.child_by_field_name("name").or_else(|| fn_node.child(0));
            if let Some(name) = name {
                symbols.calls.push(Call {
                    name: node_text(&name, source).to_string(),
                    line: start_line(node),
                    dynamic: None,
                    receiver: None,
                });
            }
        }
        _ => {}
    }
}

fn handle_object_creation(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(type_node) = node.child_by_field_name("type") else { return };
    let type_name = if type_node.kind() == "generic_name" {
        type_node.child_by_field_name("name").or_else(|| type_node.child(0))
            .map(|n| node_text(&n, source).to_string())
    } else {
        Some(node_text(&type_node, source).to_string())
    };
    if let Some(name) = type_name {
        symbols.calls.push(Call {
            name,
            line: start_line(node),
            dynamic: None,
            receiver: None,
        });
    }
}

// ── Extended kinds helpers ──────────────────────────────────────────────────

fn extract_csharp_parameters(node: &Node, source: &[u8]) -> Vec<Definition> {
    let mut params = Vec::new();
    let params_node = node.child_by_field_name("parameters")
        .or_else(|| find_child(node, "parameter_list"));
    if let Some(params_node) = params_node {
        for i in 0..params_node.child_count() {
            if let Some(child) = params_node.child(i) {
                if child.kind() == "parameter" {
                    if let Some(name_node) = child.child_by_field_name("name") {
                        params.push(child_def(
                            node_text(&name_node, source).to_string(),
                            "parameter",
                            start_line(&child),
                        ));
                    }
                }
            }
        }
    }
    params
}

fn extract_csharp_class_fields(node: &Node, source: &[u8]) -> Vec<Definition> {
    let mut fields = Vec::new();
    let body = node.child_by_field_name("body")
        .or_else(|| find_child(node, "declaration_list"));
    let Some(body) = body else { return fields };
    for i in 0..body.child_count() {
        let Some(child) = body.child(i) else { continue };
        if child.kind() == "field_declaration" {
            collect_field_declarator_names(&child, source, &mut fields);
        }
    }
    fields
}

fn collect_field_declarator_names(field: &Node, source: &[u8], fields: &mut Vec<Definition>) {
    for j in 0..field.child_count() {
        let Some(decl) = field.child(j) else { continue };
        if decl.kind() != "variable_declaration" { continue; }
        for k in 0..decl.child_count() {
            let Some(declarator) = decl.child(k) else { continue };
            if declarator.kind() != "variable_declarator" { continue; }
            let name_node = declarator.child_by_field_name("name")
                .or_else(|| declarator.child(0));
            let Some(name_node) = name_node else { continue };
            if name_node.kind() == "identifier" {
                fields.push(child_def(
                    node_text(&name_node, source).to_string(),
                    "property",
                    start_line(field),
                ));
            }
        }
    }
}

fn extract_csharp_enum_members(node: &Node, source: &[u8]) -> Vec<Definition> {
    let mut members = Vec::new();
    let body = node.child_by_field_name("body")
        .or_else(|| find_child(node, "enum_member_declaration_list"));
    if let Some(body) = body {
        for i in 0..body.child_count() {
            if let Some(child) = body.child(i) {
                if child.kind() == "enum_member_declaration" {
                    if let Some(name_node) = child.child_by_field_name("name")
                        .or_else(|| child.child(0))
                    {
                        if name_node.kind() == "identifier" {
                            members.push(child_def(
                                node_text(&name_node, source).to_string(),
                                "constant",
                                start_line(&child),
                            ));
                        }
                    }
                }
            }
        }
    }
    members
}

// ── Existing helpers ────────────────────────────────────────────────────────

/// Post-walk pass: reclassify `extends` entries as `implements` when the target
/// is a known interface in the same file. At extraction time we cannot distinguish
/// base classes from interfaces in the base_list, so we fix it up here using the
/// definitions collected during the walk.
///
/// Known limitation (mirrors the WASM `reclassifyCSharpImplements`): this pass only
/// inspects interfaces declared in the current file. Classes that implement an
/// interface declared in a different file will keep the entry as `extends` and will
/// surface as inheritance edges rather than implementation edges. Cross-file
/// reclassification would require a project-wide post-build pass.
fn reclassify_csharp_implements(symbols: &mut FileSymbols) {
    let interfaces: HashSet<String> = symbols
        .definitions
        .iter()
        .filter(|d| d.kind == "interface")
        .map(|d| d.name.clone())
        .collect();
    for cls in symbols.classes.iter_mut() {
        if let Some(ext) = cls.extends.as_ref() {
            if interfaces.contains(ext) {
                cls.implements = cls.extends.take();
            }
        }
    }
}

fn extract_csharp_base_types(
    node: &Node,
    class_name: &str,
    source: &[u8],
    symbols: &mut FileSymbols,
) {
    // tree-sitter-c-sharp exposes base_list as a child node type, not a field,
    // so child_by_field_name("bases") returns None — fall back to find_child.
    let base_list = node
        .child_by_field_name("bases")
        .or_else(|| find_child(node, "base_list"));
    let base_list = match base_list {
        Some(bl) => bl,
        None => return,
    };

    for i in 0..base_list.child_count() {
        if let Some(child) = base_list.child(i) {
            match child.kind() {
                "identifier" | "qualified_name" => {
                    symbols.classes.push(ClassRelation {
                        name: class_name.to_string(),
                        extends: Some(node_text(&child, source).to_string()),
                        implements: None,
                        line: start_line(node),
                    });
                }
                "generic_name" => {
                    let name = child
                        .child_by_field_name("name")
                        .or_else(|| child.child(0));
                    if let Some(name) = name {
                        symbols.classes.push(ClassRelation {
                            name: class_name.to_string(),
                            extends: Some(node_text(&name, source).to_string()),
                            implements: None,
                            line: start_line(node),
                        });
                    }
                }
                _ => {}
            }
        }
    }
}

// ── Type map extraction ─────────────────────────────────────────────────────

fn extract_csharp_type_name<'a>(type_node: &Node<'a>, source: &'a [u8]) -> Option<&'a str> {
    match type_node.kind() {
        "identifier" | "qualified_name" => Some(node_text(type_node, source)),
        "predefined_type" => None, // skip int, string, etc.
        "generic_name" => type_node.child(0).map(|n| node_text(&n, source)),
        "nullable_type" => {
            type_node.child(0).and_then(|inner| extract_csharp_type_name(&inner, source))
        }
        _ => None,
    }
}

/// Handle `variable_declaration` nodes: infer type from `var` (implicit_type) via
/// constructor initializer, or use the explicit type annotation for each declarator.
fn handle_csharp_var_decl_type_map(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let type_node = match node.child_by_field_name("type").or_else(|| node.child(0)) {
        Some(n) => n,
        None => return,
    };
    if type_node.kind() == "implicit_type" {
        // var x = new Foo() — infer type from object_creation_expression initializer
        for i in 0..node.child_count() {
            let Some(declarator) = node.child(i) else { continue };
            if declarator.kind() != "variable_declarator" { continue; }
            let name_node = declarator.child_by_field_name("name")
                .or_else(|| declarator.child(0));
            let Some(name_node) = name_node else { continue };
            if name_node.kind() != "identifier" { continue; }
            let Some(obj_creation) = find_child(&declarator, "object_creation_expression") else { continue };
            let Some(ctor_type_node) = obj_creation.child_by_field_name("type") else { continue };
            if let Some(ctor_type) = extract_csharp_type_name(&ctor_type_node, source) {
                symbols.type_map.push(TypeMapEntry {
                    name: node_text(&name_node, source).to_string(),
                    type_name: ctor_type.to_string(),
                    confidence: 1.0,
                });
            }
        }
    } else if type_node.kind() != "var_keyword" {
        // Explicit type: T x = ...; — emit an entry for each variable_declarator
        if let Some(type_name) = extract_csharp_type_name(&type_node, source) {
            for i in 0..node.child_count() {
                let Some(child) = node.child(i) else { continue };
                if child.kind() != "variable_declarator" { continue; }
                let name_node = child.child_by_field_name("name").or_else(|| child.child(0));
                if let Some(name_node) = name_node {
                    if name_node.kind() == "identifier" {
                        symbols.type_map.push(TypeMapEntry {
                            name: node_text(&name_node, source).to_string(),
                            type_name: type_name.to_string(),
                            confidence: 0.9,
                        });
                    }
                }
            }
        }
    }
}

/// Handle `parameter` nodes: emit a type_map entry for typed method/constructor parameters.
fn handle_csharp_param_type_map(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(type_node) = node.child_by_field_name("type") else { return };
    let Some(type_name) = extract_csharp_type_name(&type_node, source) else { return };
    let Some(name_node) = node.child_by_field_name("name") else { return };
    symbols.type_map.push(TypeMapEntry {
        name: node_text(&name_node, source).to_string(),
        type_name: type_name.to_string(),
        confidence: 0.9,
    });
}

fn match_csharp_type_map(node: &Node, source: &[u8], symbols: &mut FileSymbols, _depth: usize) {
    match node.kind() {
        "variable_declaration" => handle_csharp_var_decl_type_map(node, source, symbols),
        "parameter" => handle_csharp_param_type_map(node, source, symbols),
        _ => {}
    }
}
